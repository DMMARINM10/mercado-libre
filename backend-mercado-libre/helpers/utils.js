const numberSeparation = (number) => {
    const integer = Math.floor(number);
    const decimals = Number((number - integer).toFixed(2));
    return {
        integer,
        decimals,
    }
}

const arrayCategories = (data) => {
    const availableFilters = data.available_filters.find((filter) => filter.id === 'category');
    if(availableFilters) {
        const categories = availableFilters?.values?.sort((a, b) => b.results - a.results).map((value) => value.name);
        return categories;
    }
    const filter = data?.filters?.find((filter) => filter?.id === 'category');
    const categories = filter?.values[0]?.path_from_root?.map((cat) => cat.name)
    return categories;
}

module.exports = {
    numberSeparation,
    arrayCategories,
}