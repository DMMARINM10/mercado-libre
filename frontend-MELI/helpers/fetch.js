export const fetcher = (url) => {
    return fetch(url)
    .then((res) => {
        if(!res.ok) {
            throw new Error()
        }
        return res.json()
    });
}

export const getItem = async (id) => {
    const url = `${process.env.NEXT_PUBLIC_API_URL}/${id}`
    try {
        const resp = await fetch(url)
        if(!resp.ok) {
            return {
                error: true,
                data: null,
            }
        }
        const data = await resp.json()
        return {
            data,
            error: false
        }
    } catch (error) {
        console.error(error)
        return {
            error: true,
            data: null,
        }
    }
}