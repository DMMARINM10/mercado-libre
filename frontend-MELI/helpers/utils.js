export const formatNumber = (number) => {
    return number.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ".");
}

export const decimalFormat = (number) => {
    const decimalString = number?.toString()
    const result = decimalString?.includes('.') 
        ? decimalString.substring(2, decimalString.length) 
        : '00'
    return result?.length !== 2 ? result + '0' : result
}