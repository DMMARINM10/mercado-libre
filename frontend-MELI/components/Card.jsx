import Image from "next/image"
import Link from "next/link"
import { formatNumber } from "@/helpers/utils"
import styles from '@/styles/components/card.module.scss'
import shippingIcon from '@/assets/shipping-icon.png'

const Card = ({ id, picture, amount, freeShipping, title, seller }) => {
  return (
    <Link href={`/items/${id}`} className={styles['item-card']} key={id}>
      <Image className={styles['item-picture']} width={180} height={180} src={picture} alt={`picture-item-${id}`} />
      <span className={styles['item-metadata']}>
        <p>$ {formatNumber(amount)} {
          freeShipping && <Image height={16} src={shippingIcon} alt="free shipping icon"/>
        }</p>
        <h2>{title}</h2>
      </span>
      <p className={styles['item-seller']}>{seller}</p>
    </Link>
  )
}

export default Card
