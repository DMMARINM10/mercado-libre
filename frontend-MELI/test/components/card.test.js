import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import Card from '@/components/Card';
import { formatNumber } from '@/helpers/utils';
import shippingIcon from '@/assets/shipping-icon.png';

// Mock next/link and next/image
jest.mock('next/link', () => {
  return ({ children, href }) => <a href={href}>{children}</a>;
});

jest.mock('next/image', () => {
  return ({ src, alt, ...props }) => <img src={src} alt={alt} {...props} />;
});

describe('Card component', () => {
  const props = {
    id: '1',
    picture: 'http://random-url.jpg',
    amount: 1000,
    freeShipping: true,
    title: 'Test Title',
    seller: 'Test Seller'
  };

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render the card with correct data', () => {
    render(<Card {...props} />);

    const link = screen.getByRole('link');
    expect(link).toHaveAttribute('href', `/items/${props.id}`);

    const image = screen.getByAltText(`picture-item-${props.id}`);
    expect(image).toBeInTheDocument();

    const price = screen.getByText(`$ ${formatNumber(props.amount)}`);
    expect(price).toBeInTheDocument();

    const shippingImage = screen.getByAltText('free shipping icon');
    expect(shippingImage).toBeInTheDocument();

    const title = screen.getByText(props.title);
    expect(title).toBeInTheDocument();

    const seller = screen.getByText(props.seller);
    expect(seller).toBeInTheDocument();
  });

  it('should not render the free shipping icon if freeShipping is false', () => {
    const modifiedProps = { ...props, freeShipping: false };
    render(<Card {...modifiedProps} />);

    const shippingImage = screen.queryByAltText('free shipping icon');
    expect(shippingImage).not.toBeInTheDocument();
  });


  it('should match the snapshot', () => {
    const { asFragment } = render(<Card {...props}/>);
    expect(asFragment()).toMatchSnapshot();
  });
});
