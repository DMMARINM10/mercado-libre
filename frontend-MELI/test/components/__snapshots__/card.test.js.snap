// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component should match the snapshot 1`] = `
<DocumentFragment>
  <a
    href="/items/1"
  >
    <img
      alt="picture-item-1"
      class="item-picture"
      height="180"
      src="http://random-url.jpg"
      width="180"
    />
    <span
      class="item-metadata"
    >
      <p>
        $ 1.000 
        <img
          alt="free shipping icon"
          height="16"
          src="[object Object]"
        />
      </p>
      <h2>
        Test Title
      </h2>
    </span>
    <p
      class="item-seller"
    >
      Test Seller
    </p>
  </a>
</DocumentFragment>
`;
